app/tests/integration/nimbus/test_e2e_create_experiment.py

```diff
@@ -137,3 +137,113 @@ def test_create_new_experiment_remote_settings_timeout(
+        except AssertionError:
+            time.sleep(2)
+            selenium.refresh()
+        else:
+            break
+    # Check it's live
+    home = HomePage(selenium, base_url).wait_for_page_to_load()
+    live_experiments = home.tables[0]
+    for item in live_experiments.experiments:
+        if default_data.public_name in item.text:
+            item.click()
+            break
+    summary_page = SummaryPage(selenium, base_url).wait_for_page_to_load()
+    summary_page.end_experiment()
+    summary_page.approve()
+    perform_kinto_action(selenium, base_url, "approve")
+    # refresh page until kinto updates
+    for _ in range(45):
+        selenium.get(f"{base_url}/{request.node.name}")
+        summary_page = SummaryPage(selenium, base_url).wait_for_page_to_load()
+        if "Complete" in summary_page.experiment_status:
+            break
+        else:
+            time.sleep(2)
+            selenium.refresh()
+
+
+def test_end_experiment_and_reject_end(
+    selenium, base_url, default_data, create_experiment, request, perform_kinto_action
+):
+    default_data.public_name = request.node.name
+
+    selenium.get(base_url)
+    home = HomePage(selenium, base_url).wait_for_page_to_load()
+    current_experiments = None
+    try:
+        current_experiments = len(home.tables[0].experiments)
+    except TimeoutException:
+        current_experiments = 0
+    review = create_experiment(selenium, home, default_data)
+    review.launch_without_preview.click()
+    review.request_review.click_launch_checkboxes()
+    review.request_review.request_launch_button.click()
+    review.approve()
+
+    perform_kinto_action(selenium, base_url, "approve")
+
+    selenium.get(base_url)
+    # refresh until the experiment shows up
+    for attempt in range(45):
+        try:
+            home = HomePage(selenium, base_url).wait_for_page_to_load()
+            new_experiments = len(home.tables[0].experiments)
+            assert new_experiments > current_experiments
+        except AssertionError:
+            time.sleep(2)
+            selenium.refresh()
+        else:
+            break
+    # Check it's live
+    home = HomePage(selenium, base_url).wait_for_page_to_load()
+    live_experiments = home.tables[0]
+    for item in live_experiments.experiments:
+        if default_data.public_name in item.text:
+            item.click()
+            break
+    summary_page = SummaryPage(selenium, base_url).wait_for_page_to_load()
+    summary_page.end_experiment()
+    summary_page.approve()
+    perform_kinto_action(selenium, base_url, "reject")
+    # refresh page until kinto updates
+    for _ in range(45):
+        selenium.get(f"{base_url}/{request.node.name}")
+        summary_page = SummaryPage(selenium, base_url).wait_for_page_to_load()
+        try:
+            summary_page.rejected_text
+        except NoSuchElementException:
+            time.sleep(2)
+            selenium.refresh()
+        else:
+            break
```

Comment:
Interestingly, these two new tests look pretty similar, with a couple small diffs:

![a_txt_↔_b_txt_—_tmp](https://user-images.githubusercontent.com/557895/129980691-c6b4f760-8786-4c8b-b9f1-7e3670994c7a.png)

Not sure if they're worth trying to DRY out, or if we leave that for a future over-optimization.

Reply:
Ha, this was something I learnt the other day via Twitter. VSCode as a diff tool: <kbd>code --diff a.txt b.txt</kbd>

Reply:
Yeah, they are very similar but different enough. No clean way to parameterize them

Reply:
Yeah, since it only looked like 4-6 lines were different I didnt know if it was just adding one more param for `kinto_action` which is either "approve" or "reject" (which gets passed to `perform_kinto_action()`) and then you'd probably need another `if` statement at the bottom to say:

```python
if kinto_action == "approve":
  if "Complete" in summary_page.experiment_status:
    break;
  else:
    time.sleep(2)
    selenium.refresh()
elif kinto_action == "reject":
  try:
    assert(summary_page.rejected_text)
  except NoSuchElementException:
    time.sleep(2)
    selenium.refresh()
  else:
    break;
else:
  # Unknown kinto_action. Blow up or something.
```

But maybe thats a terrible idea. Since it works, maybe we don't tempt fate by trying to be too clever with one method to rule them all.